design-proposals/klip-7.md

```diff
@@ -0,0 +1,176 @@
+    this.properties = Objects.requireNonNull(properties, "properties");
+    this.source = Objects.requireNonNull(source, "source");
+    this.selectExpressions = ImmutableList.copyOf(selectExpressions);
+  }
+
+  public List<SelectExpression> getSelectExpressions() {
+    return selectExpressions;
+  }
+
+  @Override
+  public List<ExecutionStep> getSources() {
+    return Collections.singletonList(source);
+  }
+
+  @Override
+  public KStream build(final KStream input) {
+      final Selection selection = new Selection(selectExpressions);
+      return kstream.mapValues(selection.getSelectValueMapper()),
+  }
+}
+```
+
+I've written up a basic implementation of Execution Plans which includes all the Execution Steps we'd need, and can be found here (https://github.com/confluentinc/ksql/blob/ksql-execution-plans/ksql-common/src/main/java/io/confluent/ksql/util/KsqlConfig.java) (steps are under ksql-engine/src/main/java/io/confluent/ksql/structured/execution):
+
+#### Serializability
+The Execution Plan needs to be JSON-serializable. This way, we can persist the plan generated for each case in our Query Translation tests, and verify that it is kept the same across versions.
+
+#### Versioning
+We may need to version Execution Steps. Remember that a step is meant to represent a basic stream processing operation performed by a call to Kafka Streams. We may want to evolve how we perform a given operation in a way that is incompatible across versions. If a developer needs to modify a step in an incompatible way, they can do so by writing a new step type and advancing the version.
+
+### Longer Term
+The above proposal should cover the existing gaps in checking engine compatibility across versions. However it doesn't address many of the issues identified in the Background/Motivation section of this document:
+
+- Tests are always incomplete. There are always going to be gaps in what the tests cover, leaving us vulnerable to queries that are not verified for compatibility.
+- The tests take a long time to run. Currently there are 473 test cases that must be repeated for every previous version of KSQL. Today this takes about a minute. This is going to grow as we add language features and support CCloud, where we want to be able to release much more frequently.
+- The usage of compatibility-breaking configs adds a lot of complexity to the codebase. We have to maintain ugly code in the engine that switches on these configs to preserve legacy behaviour.
+
+To solve the above I propose that when accepting statements, in addition to persisting the original statements and configs, we persist the Execution Plan as well. Then, when executing queries, we can build the final streams topologies from the persisted Execution Plan instead of rebuilding the queries from statements every time. This way, we don't need compatibility verification, and don't need to worry about query compatibility in most of the engine code.
```

Comment:
> This way, we don't need compatibility verification

Can you clarify what the proposal for the existing query translation tests is in this long term solution? It sounds like the proposal is to do away with the topology and schema checks, leaving only input/output verification. Would we only perform input/output verification on the `master` branch, or is it also worth validating how old configs potentially affect inputs/outputs?

Reply:
> Can you clarify what the proposal for the existing query translation tests is in this long term solution? It sounds like the proposal is to do away with the topology and schema checks, leaving only input/output verification.

Yeah that's correct.

> Would we only perform input/output verification on the master branch, or is it also worth validating how old configs potentially affect inputs/outputs?

Not sure what you mean here. Can you elaborate? Do you mean the existing compatibility verification tests? We'd still need those until at least the next major version bump.